src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java

```diff
@@ -261,21 +263,34 @@ private String findTargetProcessTmpDirectory(long pid, long ns_pid) throws Attac
          * 4. Caller and target processes share neither PID namespace nor root filesystem (host to container)
          *
          * if target is elevated, we cant use /proc/<pid>/... so we have to fallback to /tmp, but that may not be shared
-         * with the target/attachee process, we can try, except in the case where the ns_pid also exists in this pid ns
-         * which is ambiguous, if we share /tmp with the intended target, the attach will succeed, if we do not,
-         * then we will potentially attempt to attach to some arbitrary process with the same pid (in this pid ns)
-         * as that of the intended target (in its * pid ns).
-         *
-         * so in that case we should prehaps throw - or risk sending SIGQUIT to some arbitrary process... which could kill it
+         * with the target/attachee process, so we should check whether /tmp on both is same. This method would throw
+         * AttachNotSupportedException if they are different because we cannot make a connection with target VM.
          *
-         * however we can also check the target pid's signal masks to see if it catches SIGQUIT and only do so if in
+         * In addition, we can also check the target pid's signal masks to see if it catches SIGQUIT and only do so if in
          * fact it does ... this reduces the risk of killing an innocent process in the current ns as opposed to
          * attaching to the actual target JVM ... c.f: checkCatchesAndSendQuitTo() below.
-         *
-         * note that if pid == ns_pid we are in a shared pid ns with the target and may (potentially) share /tmp
          */
 
-        return (Files.isWritable(procPidRoot) ? procPidRoot : TMPDIR).toString();
+        try {
+            if (Files.isWritable(tmpOnProcPidRoot)) {
+                return tmpOnProcPidRoot.toString();
+            } else if (Files.isSameFile(tmpOnProcPidRoot, TMPDIR)) {
+                return TMPDIR.toString();
+            } else {
+                boolean found = MonitoredHost.getMonitoredHost("//localhost")
+                                             .activeVms()
+                                             .stream()
+                                             .anyMatch(i -> pid == i.intValue());
+                if (found) {
+                    // We can use /tmp because target process is on same host
+                    return TMPDIR.toString();
+                } else {
+                    throw new AttachNotSupportedException("Unable to access the filesystem of the target process");
+                }
+            }
+        } catch (Exception e) {
+            throw new AttachNotSupportedException("Unable to access the filesystem of the target process", e);
+        }
```

Comment:
Not sure if this stays like it is, but if it does, we can probably avoid wrapping `AttachNotSupportedException`:

```suggestion
        } catch (AttachNotSupportedException e) {
            throw e;
        } catch (Exception e) {
            throw new AttachNotSupportedException("Unable to access the filesystem of the target process", e);
        }
```